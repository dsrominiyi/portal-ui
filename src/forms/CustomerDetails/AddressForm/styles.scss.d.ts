export const addressContainer: string;
export const validationError: string;
