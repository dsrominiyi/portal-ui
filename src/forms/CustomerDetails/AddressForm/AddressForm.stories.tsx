import React from 'react';
import { storiesOf } from '@storybook/react';
import AddressForm, { emptyAddress } from './index';

const stories = storiesOf('AddressForm', module);

export const formData: any = {};

const save = (data: any) => {
  Object.keys(save).forEach(itemKey => {
    formData[itemKey] = data[itemKey];
  });
};

const render = (address?: Address) => () => (
  <AddressForm initialValues={address} onChange={() => save(formData)} key="address" />
);

export const storyRenders = {
  emptyAddress: render(emptyAddress),
  noAddress: render()
};

stories.add('Empty Address', storyRenders.emptyAddress);
stories.add('No Address', storyRenders.noAddress);
