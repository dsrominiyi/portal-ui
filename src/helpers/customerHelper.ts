import { dateFieldValuesToDateString, dateStringToDateFieldValues } from './utilFuncs';
import { predefinedRules, ValidationRules } from './formValidator';

export const companyFormValidationRules: ValidationRules = {
  companyName: { ...predefinedRules.entityName, fieldLabel: 'company name', required: true },
  tradingName: { ...predefinedRules.entityName, fieldLabel: 'trading name' },
  typeStatus: { fieldLabel: 'company legal status', invalidValues: ['Other'], required: true },
  businessClass: { fieldLabel: 'business class', required: true },
  tradingSinceDate: { fieldLabel: 'trading since', isDate: true },
  parentCompany: { fieldLabel: 'is this company a subsidiary', required: true },
  parentCompanyName: { ...predefinedRules.entityName, fieldLabel: 'parent company name' },
  registeredCoNumber: predefinedRules.registeredCoNumber,
  contactDetails: { fieldLabel: 'contact details', required: true },
  addresses: { fieldLabel: 'addresses', required: true }
};

export const isIndividualCustomer = (customer: Customer): customer is Individual => {
  return typeof (customer as Individual).name !== 'undefined';
};

export const getIndividualFullName = ({ title, forename, middleName, surname }: IndividualName) => {
  return `${title} ${forename}${middleName && ` ${middleName}`} ${surname}`;
};

export const buildCompanyFromFormData = (data: NewCompanyData): Company => ({
  ...data,
  parentCompany: data.parentCompany || false,
  tradingSinceDate: dateFieldValuesToDateString(data.tradingSinceDate)
});

export const buildFormDataFromCompany = (data: Company): NewCompanyData => ({
  ...data,
  tradingSinceDate: data.tradingSinceDate
    ? dateStringToDateFieldValues(data.tradingSinceDate)
    : { dd: '', mm: '', yyyy: '' }
});
