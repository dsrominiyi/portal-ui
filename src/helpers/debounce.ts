/**
 * Prevents an action from being called more that once every `wait` milliseconds
 * The initial triggering of the action will be delayed by `wait` milliseconds if `immediate` = false
 *
 * e.g. <button onClick={event => debounce(handleClick, 200)(event)} />
 */

let timeout: number | null;

const debounce = (action: (...args: any[]) => void, wait = 300, immediate = true) => (
  ...args: any[]
) => {
  const later = () => {
    timeout = null;
    if (!immediate) {
      action(...args);
    }
  };
  const callNow = immediate && !timeout;

  if (timeout) {
    clearTimeout(timeout);
  }

  timeout = (setTimeout as Window['setTimeout'])(later, wait);

  if (callNow) {
    action(...args);
  }
};

export default debounce;
