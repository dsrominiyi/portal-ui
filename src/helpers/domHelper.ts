import debounce from './debounce';

export const preventBodyScroll = () => {
  const { scrollY } = window;
  document.body.style.position = 'fixed';
  document.body.style.top = document.body.style.top || `-${scrollY}px`;
};

export const allowBodyScroll = () => {
  const scrollY = document.body.style.top;
  document.body.style.position = '';
  document.body.style.top = '';
  return scrollY && window.scrollTo(0, -parseInt(scrollY, 10));
};

export const triggerOnEnter = (action: () => void, debounceWait?: number) => (
  event: React.KeyboardEvent<Element | HTMLElement>
) => {
  if (event.key === 'Enter') {
    debounce(action, debounceWait)();
  }
};
