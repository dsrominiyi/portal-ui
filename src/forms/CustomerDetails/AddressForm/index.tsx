import React, { ChangeEvent, FunctionComponent, useState, useEffect } from 'react';

import SubQuestion from '../../../components/SubQuestion';
import ValidationErrorsNotification from '../../common/ValidationErrorsNotification';

import {
  BatchValidationResult,
  isFormValid,
  predefinedRules,
  validateForm,
  ValidationRules,
  ValidationStateArray,
  buildErrors
} from '../../../helpers/formValidator';
import {
  individualAddressTypeOptions,
  CompanyAddressTypeOptions
} from '../../../helpers/optionLists';
import { triggerOnEnter } from '../../../helpers/domHelper';
import { buildInput } from '../../../helpers/formHelper';

import formStyles from '../../styles.scss';

interface FormConfig {
  customerType: CustomerType;
}

type UseStateArray = [Address, React.Dispatch<React.SetStateAction<Address>>];

export const emptyAddress: Address = {
  type: undefined,
  organisation: '',
  buildingName: '',
  subBuildingName: '',
  buildingNumber: '',
  thoroughfare: '',
  subThoroughfare: '',
  locality: '',
  town: '',
  county: '',
  postcode: ''
};

const validationRules: ValidationRules = {
  type: { fieldLabel: 'address type' },
  organisation: { fieldLabel: 'organisation', required: false, maxLength: 50 },
  buildingName: { fieldLabel: 'building name', required: true, maxLength: 50 },
  subBuildingName: { fieldLabel: 'subBuildingName', required: false },
  buildingNumber: { fieldLabel: 'building number', required: true, maxLength: 50 },
  thoroughfare: { fieldLabel: 'first line', required: true, maxLength: 50 },
  subThoroughfare: { fieldLabel: 'second line', required: false, maxLength: 50 },
  locality: { fieldLabel: 'locality', required: false, maxLength: 50 },
  town: { fieldLabel: 'town', required: true, maxLength: 50 },
  county: { fieldLabel: 'county', required: true, maxLength: 50 },
  postcode: { ...predefinedRules.postcode, required: true }
};

export const AddressForm: FunctionComponent<FormProps> = ({
  additionalClasses = [],
  initialValues,
  submissionAttempted: submissionAttemptedProp,
  config = {},
  onChange,
  onSubmit
}: FormProps) => {
  const [state, setState]: UseStateArray = useState((initialValues as Address) || emptyAddress);

  const [validationState, setValidationState]: ValidationStateArray = useState(
    {} as BatchValidationResult
  );
  const [submissionAttempted, setSubmissionAttempted] = useState(false);
  const [manualEntry, setManualEntry] = useState(false);

  const canSubmit = isFormValid(validationState);

  const getBuildingNameOrNumberRules = (validationAddress: Address) => {
    const buildingNameOrNumberRules = {
      buildingName: { ...validationRules.buildingName },
      buildingNumber: { ...validationRules.buildingNumber }
    };
    if (validationAddress.buildingNumber) {
      buildingNameOrNumberRules.buildingName.required = false;
      return buildingNameOrNumberRules;
    }
    if (validationAddress.buildingName) {
      buildingNameOrNumberRules.buildingNumber.required = false;
      return buildingNameOrNumberRules;
    }
    return buildingNameOrNumberRules;
  };

  useEffect(() => {
    const allValidationRules = { ...validationRules, ...getBuildingNameOrNumberRules(state) };

    validateForm(allValidationRules, { ...state }, setValidationState);
    onChange(state, canSubmit);
  }, [state]);

  useEffect(() => {
    onChange(state, canSubmit);
  }, [validationState]);

  const getTextOnChange = (fieldName: keyof Address) => {
    return (event: ChangeEvent<HTMLInputElement>) => {
      let { value } = event.target;
      if (fieldName === 'postcode') {
        value = value.toUpperCase();
      }
      setState({ ...state, [fieldName]: value });
    };
  };

  const toggleManualEntry = () => setManualEntry(!manualEntry);

  const submit = () => {
    setSubmissionAttempted(true);
    return canSubmit && onSubmit && onSubmit();
  };

  const { customerType } = config as FormConfig;
  const hasAttemptedToSubmit = submissionAttemptedProp || submissionAttempted;
  const formErrors = buildErrors(validationState, hasAttemptedToSubmit);

  return (
    <div
      className={[formStyles.formSection, additionalClasses].join(' ')}
      onKeyUp={triggerOnEnter(submit)}
      test-id="new-address-form"
    >
      {hasAttemptedToSubmit && !canSubmit && <ValidationErrorsNotification />}
      {buildInput(
        {
          type: InputType.Dropdown,
          label: 'Type',
          options:
            customerType === 'Individual'
              ? individualAddressTypeOptions
              : CompanyAddressTypeOptions,
          value: state.type,
          onChange: (option: AddressType) => {
            setState({ ...state, type: option });
          },
          attributes: { 'test-id': 'select-type' }
        },
        formErrors.type
      )}
      {buildInput(
        {
          type: InputType.Text,
          label: 'Building name/number',
          value: state.buildingNumber,
          onChange: getTextOnChange('buildingNumber'),
          attributes: { 'test-id': 'building-name-number' }
        },
        formErrors.buildingNumber
      )}
      {buildInput(
        {
          type: InputType.Lookup,
          label: 'Postcode',
          value: state.postcode,
          buttonLabel: 'Find Address',
          secondaryButtonLabel: 'Enter Manually',
          onChange: getTextOnChange('postcode'),
          secondaryAction: toggleManualEntry,
          attributes: { 'test-id': 'postcode-lookup' }
        },
        formErrors.postcode
      )}
      <SubQuestion
        visible={manualEntry}
        items={[
          buildInput(
            {
              type: InputType.Text,
              label: 'Organisation',
              value: state.organisation,
              onChange: getTextOnChange('organisation'),
              attributes: { 'test-id': 'organisation' }
            },
            formErrors.organisation
          ),
          buildInput(
            {
              type: InputType.Text,
              label: 'Building name',
              value: state.buildingName,
              onChange: getTextOnChange('buildingName'),
              attributes: { 'test-id': 'building-name' }
            },
            formErrors.buildingName
          ),
          buildInput(
            {
              type: InputType.Text,
              label: 'Building number',
              value: state.buildingNumber,
              onChange: getTextOnChange('buildingNumber'),
              attributes: { 'test-id': 'building-number' }
            },
            formErrors.buildingNumber
          ),
          buildInput(
            {
              type: InputType.Text,
              label: 'First Line',
              value: state.thoroughfare,
              onChange: getTextOnChange('thoroughfare'),
              attributes: { 'test-id': 'line-1' }
            },
            formErrors.thoroughfare
          ),
          buildInput(
            {
              type: InputType.Text,
              label: 'Second Line',
              value: state.subThoroughfare,
              onChange: getTextOnChange('subThoroughfare'),
              attributes: { 'test-id': 'line-2' }
            },
            formErrors.subThoroughfare
          ),
          buildInput(
            {
              type: InputType.Text,
              label: 'Town',
              value: state.town,
              onChange: getTextOnChange('town'),
              attributes: { 'test-id': 'town' }
            },
            formErrors.town
          ),
          buildInput(
            {
              type: InputType.Text,
              label: 'County',
              value: state.county,
              onChange: getTextOnChange('county'),
              attributes: { 'test-id': 'county' }
            },
            formErrors.county
          ),
          buildInput(
            {
              type: InputType.Text,
              label: 'Postcode',
              value: state.postcode,
              onChange: getTextOnChange('postcode'),
              attributes: { 'test-id': 'postcode' }
            },
            formErrors.postcode
          )
        ]}
      />
    </div>
  );
};

export default AddressForm;
