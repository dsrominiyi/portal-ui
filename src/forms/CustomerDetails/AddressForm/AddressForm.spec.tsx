import { mount, shallow } from 'enzyme';
import React from 'react';
import { act } from 'react-dom/test-utils';

import AddressForm, { emptyAddress } from './index';
import { isFormValid, validateForm, buildErrors } from '../../../helpers/formValidator';

jest.mock('../../../helpers/formValidator');

let form: any;
const onSubmit = jest.fn().mockReturnValue(true);
const onChange = jest.fn();
(isFormValid as jest.Mock).mockReturnValue(true);
(buildErrors as jest.Mock).mockReturnValue({
  organisation: {},
  buildingName: {},
  buildingNumber: {},
  thoroughfare: {},
  subThoroughfare: {},
  town: {},
  county: {},
  postcode: {}
});

const getField = (fieldIndex: number) => {
  return (form
    .find('QuestionBox')
    .at(fieldIndex)
    .props() as any).inputFields[0];
};

const getSubField = (fieldIndex: number) => {
  return (form
    .find('SubQuestion')
    .first()
    .props() as any).items[fieldIndex].props.inputFields[0];
};

const address = { ...emptyAddress, postcode: 'A12 3BC' };

describe('AddressForm', () => {
  beforeEach(() => {
    jest.clearAllMocks();
    form = shallow(
      <AddressForm
        config={{ customerType: 'Individual' }}
        onChange={onChange}
        onSubmit={onSubmit}
      />
    );
  });

  describe('snapshots', () => {
    test('AddressForm for individual matches snapshot', () => {
      form.setProps({
        config: { customerType: 'Company' }
      });

      expect(form).toMatchSnapshot();
    });

    test('AddressForm for company matches snapshot', () => {
      expect(form).toMatchSnapshot();
    });

    test('AddressForm with initial values matches snapshot', () => {
      form.setProps({
        initialValues: address
      });

      expect(form).toMatchSnapshot();
    });

    test('SubsidiaryForm without form errors matches snapshot', () => {
      (buildErrors as jest.Mock).mockReturnValueOnce({});
      form = shallow(<AddressForm onChange={onChange} onSubmit={onSubmit} />);

      expect(form).toMatchSnapshot();
    });
  });

  describe('functionality', () => {
    const getMountedForm = () => {
      return mount(
        <AddressForm
          initialValues={address}
          onChange={onChange}
          onSubmit={onSubmit}
          key="address"
        />
      );
    };

    test('clicking the manual entry button reveals the manual entry form', () => {
      const postcodeLookup = getField(2);
      postcodeLookup.props.secondaryAction();
      form.update();
      const manualEntryForm = form.find('SubQuestion').first();
      expect(manualEntryForm.props().visible).toBe(true);
    });

    test('selecting a type updates the type state value', () => {
      act(() => {
        getField(0).props.onChange('residential');
      });
      form.update();

      expect(getField(0).props.selected).toBe('residential');
    });

    test('updating the town field updates the town state value', () => {
      const input = getSubField(5);
      act(() => {
        (input.props as any).onChange({ target: { value: 'anytown' } });
      });
      form.update();

      expect(getSubField(5).props.value).toBe('anytown');
    });

    test('updating the postcode field updates the postcode state value', () => {
      getField(2).props.onChange({ target: { value: 'TE1 5ST' } });
      form.update();

      expect(getField(2).props.value).toBe('TE1 5ST');
    });

    test('onSubmit is called when the Enter is pressed', () => {
      (isFormValid as jest.Mock).mockReturnValueOnce(true);
      form.simulate('keyup', { key: 'Enter' });

      expect(onSubmit).toHaveBeenCalled();
    });

    test('onSubmit is not called when the Enter is pressed but the form is invalid', () => {
      (isFormValid as jest.Mock).mockReturnValueOnce(false);
      form.simulate('keyup', { key: 'Enter' });

      expect(onSubmit).not.toHaveBeenCalled();
    });

    test('ValidationErrorsNotification is rendered when the Enter is pressed but the form is invalid', () => {
      (isFormValid as jest.Mock).mockReturnValueOnce(false);
      form.setProps({ submissionAttempted: true });
      form.simulate('keyup', { key: 'Enter' });

      expect(form.find('ValidationErrorsNotification').length).toBe(1);
    });

    test('entering a building name makes building number optional', () => {
      const mountedForm = getMountedForm();
      const buildingName = (mountedForm
        .find('SubQuestion')
        .first()
        .props() as any).items[1].props.inputFields[0];
      act(() => {
        buildingName.props.onChange({ target: { value: 'Any House' } });
      });
      mountedForm.update();
      mountedForm.simulate('keyup', { key: 'Enter' });

      expect(validateForm).toHaveBeenLastCalledWith(
        expect.objectContaining({
          buildingName: expect.objectContaining({ required: true }),
          buildingNumber: expect.objectContaining({ required: false })
        }),
        expect.anything(),
        expect.anything()
      );
    });

    test('entering a building number makes building name optional', () => {
      const mountedForm = getMountedForm();
      const buildingNumber = (mountedForm
        .find('SubQuestion')
        .first()
        .props() as any).items[2].props.inputFields[0];
      act(() => {
        buildingNumber.props.onChange({ target: { value: 'Any House' } });
      });
      mountedForm.update();
      mountedForm.simulate('keyup', { key: 'Enter' });

      expect(validateForm).toHaveBeenLastCalledWith(
        expect.objectContaining({
          buildingName: expect.objectContaining({ required: false }),
          buildingNumber: expect.objectContaining({ required: true })
        }),
        expect.anything(),
        expect.anything()
      );
    });

    test('onChange is called with the form state when a field is updated', () => {
      const postcode = 'SD1 4VB';

      act(() => {
        (getMountedForm()
          .find('Lookup')
          .at(0)
          .props() as any).onChange({ target: { value: postcode } });
      });

      expect(onChange).toHaveBeenCalledWith({ ...emptyAddress, postcode }, true);
    });

    test('postcode lookup should convert everything to uppercase', () => {
      getField(2).props.onChange({ target: { value: 'te1 5st' } });
      form.update();

      expect(getField(2).props.value).toBe('TE1 5ST');
    });
  });
});
