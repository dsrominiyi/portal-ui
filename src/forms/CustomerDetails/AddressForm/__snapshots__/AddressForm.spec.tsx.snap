// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddressForm snapshots AddressForm for company matches snapshot 1`] = `
<div
  className="formSection "
  onKeyUp={[Function]}
  test-id="new-address-form"
>
  <QuestionBox
    inputFields={
      Array [
        <Dropdown
          attributes={
            Object {
              "test-id": "select-type",
            }
          }
          label="Type"
          onChange={[Function]}
          options={
            Array [
              "Home",
              "Correspondence",
            ]
          }
        />,
      ]
    }
    key="Type"
    text=""
  />
  <QuestionBox
    inputFields={
      Array [
        <TextField
          additionalClasses={
            Array [
              "formTextField",
            ]
          }
          attributes={
            Object {
              "test-id": "building-name-number",
            }
          }
          label="Building name/number"
          onChange={[Function]}
          value=""
        />,
      ]
    }
    key="Building name/number"
    text=""
  />
  <QuestionBox
    inputFields={
      Array [
        <Lookup
          attributes={
            Object {
              "test-id": "postcode-lookup",
            }
          }
          buttonLabel="Find Address"
          label="Postcode"
          onChange={[Function]}
          secondaryAction={[Function]}
          secondaryButtonLabel="Enter Manually"
          value=""
        />,
      ]
    }
    key="Postcode"
    text=""
  />
  <SubQuestion
    items={
      Array [
        <QuestionBox
          inputFields={
            Array [
              <TextField
                additionalClasses={
                  Array [
                    "formTextField",
                  ]
                }
                attributes={
                  Object {
                    "test-id": "organisation",
                  }
                }
                label="Organisation"
                onChange={[Function]}
                value=""
              />,
            ]
          }
          text=""
        />,
        <QuestionBox
          inputFields={
            Array [
              <TextField
                additionalClasses={
                  Array [
                    "formTextField",
                  ]
                }
                attributes={
                  Object {
                    "test-id": "building-name",
                  }
                }
                label="Building name"
                onChange={[Function]}
                value=""
              />,
            ]
          }
          text=""
        />,
        <QuestionBox
          inputFields={
            Array [
              <TextField
                additionalClasses={
                  Array [
                    "formTextField",
                  ]
                }
                attributes={
                  Object {
                    "test-id": "building-number",
                  }
                }
                label="Building number"
                onChange={[Function]}
                value=""
              />,
            ]
          }
          text=""
        />,
        <QuestionBox
          inputFields={
            Array [
              <TextField
                additionalClasses={
                  Array [
                    "formTextField",
                  ]
                }
                attributes={
                  Object {
                    "test-id": "line-1",
                  }
                }
                label="First Line"
                onChange={[Function]}
                value=""
              />,
            ]
          }
          text=""
        />,
        <QuestionBox
          inputFields={
            Array [
              <TextField
                additionalClasses={
                  Array [
                    "formTextField",
                  ]
                }
                attributes={
                  Object {
                    "test-id": "line-2",
                  }
                }
                label="Second Line"
                onChange={[Function]}
                value=""
              />,
            ]
          }
          text=""
        />,
        <QuestionBox
          inputFields={
            Array [
              <TextField
                additionalClasses={
                  Array [
                    "formTextField",
                  ]
                }
                attributes={
                  Object {
                    "test-id": "town",
                  }
                }
                label="Town"
                onChange={[Function]}
                value=""
              />,
            ]
          }
          text=""
        />,
        <QuestionBox
          inputFields={
            Array [
              <TextField
                additionalClasses={
                  Array [
                    "formTextField",
                  ]
                }
                attributes={
                  Object {
                    "test-id": "county",
                  }
                }
                label="County"
                onChange={[Function]}
                value=""
              />,
            ]
          }
          text=""
        />,
        <QuestionBox
          inputFields={
            Array [
              <TextField
                additionalClasses={
                  Array [
                    "formTextField",
                  ]
                }
                attributes={
                  Object {
                    "test-id": "postcode",
                  }
                }
                label="Postcode"
                onChange={[Function]}
                value=""
              />,
            ]
          }
          text=""
        />,
      ]
    }
    visible={false}
  />
</div>
`;

exports[`AddressForm snapshots AddressForm for individual matches snapshot 1`] = `
<div
  className="formSection "
  onKeyUp={[Function]}
  test-id="new-address-form"
>
  <QuestionBox
    inputFields={
      Array [
        <Dropdown
          attributes={
            Object {
              "test-id": "select-type",
            }
          }
          label="Type"
          onChange={[Function]}
          options={
            Array [
              "Correspondence",
              "Trading",
              "Registered",
            ]
          }
        />,
      ]
    }
    key="Type"
    text=""
  />
  <QuestionBox
    inputFields={
      Array [
        <TextField
          additionalClasses={
            Array [
              "formTextField",
            ]
          }
          attributes={
            Object {
              "test-id": "building-name-number",
            }
          }
          label="Building name/number"
          onChange={[Function]}
          value=""
        />,
      ]
    }
    key="Building name/number"
    text=""
  />
  <QuestionBox
    inputFields={
      Array [
        <Lookup
          attributes={
            Object {
              "test-id": "postcode-lookup",
            }
          }
          buttonLabel="Find Address"
          label="Postcode"
          onChange={[Function]}
          secondaryAction={[Function]}
          secondaryButtonLabel="Enter Manually"
          value=""
        />,
      ]
    }
    key="Postcode"
    text=""
  />
  <SubQuestion
    items={
      Array [
        <QuestionBox
          inputFields={
            Array [
              <TextField
                additionalClasses={
                  Array [
                    "formTextField",
                  ]
                }
                attributes={
                  Object {
                    "test-id": "organisation",
                  }
                }
                label="Organisation"
                onChange={[Function]}
                value=""
              />,
            ]
          }
          text=""
        />,
        <QuestionBox
          inputFields={
            Array [
              <TextField
                additionalClasses={
                  Array [
                    "formTextField",
                  ]
                }
                attributes={
                  Object {
                    "test-id": "building-name",
                  }
                }
                label="Building name"
                onChange={[Function]}
                value=""
              />,
            ]
          }
          text=""
        />,
        <QuestionBox
          inputFields={
            Array [
              <TextField
                additionalClasses={
                  Array [
                    "formTextField",
                  ]
                }
                attributes={
                  Object {
                    "test-id": "building-number",
                  }
                }
                label="Building number"
                onChange={[Function]}
                value=""
              />,
            ]
          }
          text=""
        />,
        <QuestionBox
          inputFields={
            Array [
              <TextField
                additionalClasses={
                  Array [
                    "formTextField",
                  ]
                }
                attributes={
                  Object {
                    "test-id": "line-1",
                  }
                }
                label="First Line"
                onChange={[Function]}
                value=""
              />,
            ]
          }
          text=""
        />,
        <QuestionBox
          inputFields={
            Array [
              <TextField
                additionalClasses={
                  Array [
                    "formTextField",
                  ]
                }
                attributes={
                  Object {
                    "test-id": "line-2",
                  }
                }
                label="Second Line"
                onChange={[Function]}
                value=""
              />,
            ]
          }
          text=""
        />,
        <QuestionBox
          inputFields={
            Array [
              <TextField
                additionalClasses={
                  Array [
                    "formTextField",
                  ]
                }
                attributes={
                  Object {
                    "test-id": "town",
                  }
                }
                label="Town"
                onChange={[Function]}
                value=""
              />,
            ]
          }
          text=""
        />,
        <QuestionBox
          inputFields={
            Array [
              <TextField
                additionalClasses={
                  Array [
                    "formTextField",
                  ]
                }
                attributes={
                  Object {
                    "test-id": "county",
                  }
                }
                label="County"
                onChange={[Function]}
                value=""
              />,
            ]
          }
          text=""
        />,
        <QuestionBox
          inputFields={
            Array [
              <TextField
                additionalClasses={
                  Array [
                    "formTextField",
                  ]
                }
                attributes={
                  Object {
                    "test-id": "postcode",
                  }
                }
                label="Postcode"
                onChange={[Function]}
                value=""
              />,
            ]
          }
          text=""
        />,
      ]
    }
    visible={false}
  />
</div>
`;

exports[`AddressForm snapshots AddressForm with initial values matches snapshot 1`] = `
<div
  className="formSection "
  onKeyUp={[Function]}
  test-id="new-address-form"
>
  <QuestionBox
    inputFields={
      Array [
        <Dropdown
          attributes={
            Object {
              "test-id": "select-type",
            }
          }
          label="Type"
          onChange={[Function]}
          options={
            Array [
              "Home",
              "Correspondence",
            ]
          }
        />,
      ]
    }
    key="Type"
    text=""
  />
  <QuestionBox
    inputFields={
      Array [
        <TextField
          additionalClasses={
            Array [
              "formTextField",
            ]
          }
          attributes={
            Object {
              "test-id": "building-name-number",
            }
          }
          label="Building name/number"
          onChange={[Function]}
          value=""
        />,
      ]
    }
    key="Building name/number"
    text=""
  />
  <QuestionBox
    inputFields={
      Array [
        <Lookup
          attributes={
            Object {
              "test-id": "postcode-lookup",
            }
          }
          buttonLabel="Find Address"
          label="Postcode"
          onChange={[Function]}
          secondaryAction={[Function]}
          secondaryButtonLabel="Enter Manually"
          value=""
        />,
      ]
    }
    key="Postcode"
    text=""
  />
  <SubQuestion
    items={
      Array [
        <QuestionBox
          inputFields={
            Array [
              <TextField
                additionalClasses={
                  Array [
                    "formTextField",
                  ]
                }
                attributes={
                  Object {
                    "test-id": "organisation",
                  }
                }
                label="Organisation"
                onChange={[Function]}
                value=""
              />,
            ]
          }
          text=""
        />,
        <QuestionBox
          inputFields={
            Array [
              <TextField
                additionalClasses={
                  Array [
                    "formTextField",
                  ]
                }
                attributes={
                  Object {
                    "test-id": "building-name",
                  }
                }
                label="Building name"
                onChange={[Function]}
                value=""
              />,
            ]
          }
          text=""
        />,
        <QuestionBox
          inputFields={
            Array [
              <TextField
                additionalClasses={
                  Array [
                    "formTextField",
                  ]
                }
                attributes={
                  Object {
                    "test-id": "building-number",
                  }
                }
                label="Building number"
                onChange={[Function]}
                value=""
              />,
            ]
          }
          text=""
        />,
        <QuestionBox
          inputFields={
            Array [
              <TextField
                additionalClasses={
                  Array [
                    "formTextField",
                  ]
                }
                attributes={
                  Object {
                    "test-id": "line-1",
                  }
                }
                label="First Line"
                onChange={[Function]}
                value=""
              />,
            ]
          }
          text=""
        />,
        <QuestionBox
          inputFields={
            Array [
              <TextField
                additionalClasses={
                  Array [
                    "formTextField",
                  ]
                }
                attributes={
                  Object {
                    "test-id": "line-2",
                  }
                }
                label="Second Line"
                onChange={[Function]}
                value=""
              />,
            ]
          }
          text=""
        />,
        <QuestionBox
          inputFields={
            Array [
              <TextField
                additionalClasses={
                  Array [
                    "formTextField",
                  ]
                }
                attributes={
                  Object {
                    "test-id": "town",
                  }
                }
                label="Town"
                onChange={[Function]}
                value=""
              />,
            ]
          }
          text=""
        />,
        <QuestionBox
          inputFields={
            Array [
              <TextField
                additionalClasses={
                  Array [
                    "formTextField",
                  ]
                }
                attributes={
                  Object {
                    "test-id": "county",
                  }
                }
                label="County"
                onChange={[Function]}
                value=""
              />,
            ]
          }
          text=""
        />,
        <QuestionBox
          inputFields={
            Array [
              <TextField
                additionalClasses={
                  Array [
                    "formTextField",
                  ]
                }
                attributes={
                  Object {
                    "test-id": "postcode",
                  }
                }
                label="Postcode"
                onChange={[Function]}
                value=""
              />,
            ]
          }
          text=""
        />,
      ]
    }
    visible={false}
  />
</div>
`;

exports[`AddressForm snapshots SubsidiaryForm without form errors matches snapshot 1`] = `
<div
  className="formSection "
  onKeyUp={[Function]}
  test-id="new-address-form"
>
  <QuestionBox
    inputFields={
      Array [
        <Dropdown
          attributes={
            Object {
              "test-id": "select-type",
            }
          }
          label="Type"
          onChange={[Function]}
          options={
            Array [
              "Correspondence",
              "Trading",
              "Registered",
            ]
          }
        />,
      ]
    }
    key="Type"
    text=""
  />
  <QuestionBox
    inputFields={
      Array [
        <TextField
          additionalClasses={
            Array [
              "formTextField",
            ]
          }
          attributes={
            Object {
              "test-id": "building-name-number",
            }
          }
          label="Building name/number"
          onChange={[Function]}
          value=""
        />,
      ]
    }
    key="Building name/number"
    text=""
  />
  <QuestionBox
    inputFields={
      Array [
        <Lookup
          attributes={
            Object {
              "test-id": "postcode-lookup",
            }
          }
          buttonLabel="Find Address"
          label="Postcode"
          onChange={[Function]}
          secondaryAction={[Function]}
          secondaryButtonLabel="Enter Manually"
          value=""
        />,
      ]
    }
    key="Postcode"
    text=""
  />
  <SubQuestion
    items={
      Array [
        <QuestionBox
          inputFields={
            Array [
              <TextField
                additionalClasses={
                  Array [
                    "formTextField",
                  ]
                }
                attributes={
                  Object {
                    "test-id": "organisation",
                  }
                }
                label="Organisation"
                onChange={[Function]}
                value=""
              />,
            ]
          }
          text=""
        />,
        <QuestionBox
          inputFields={
            Array [
              <TextField
                additionalClasses={
                  Array [
                    "formTextField",
                  ]
                }
                attributes={
                  Object {
                    "test-id": "building-name",
                  }
                }
                label="Building name"
                onChange={[Function]}
                value=""
              />,
            ]
          }
          text=""
        />,
        <QuestionBox
          inputFields={
            Array [
              <TextField
                additionalClasses={
                  Array [
                    "formTextField",
                  ]
                }
                attributes={
                  Object {
                    "test-id": "building-number",
                  }
                }
                label="Building number"
                onChange={[Function]}
                value=""
              />,
            ]
          }
          text=""
        />,
        <QuestionBox
          inputFields={
            Array [
              <TextField
                additionalClasses={
                  Array [
                    "formTextField",
                  ]
                }
                attributes={
                  Object {
                    "test-id": "line-1",
                  }
                }
                label="First Line"
                onChange={[Function]}
                value=""
              />,
            ]
          }
          text=""
        />,
        <QuestionBox
          inputFields={
            Array [
              <TextField
                additionalClasses={
                  Array [
                    "formTextField",
                  ]
                }
                attributes={
                  Object {
                    "test-id": "line-2",
                  }
                }
                label="Second Line"
                onChange={[Function]}
                value=""
              />,
            ]
          }
          text=""
        />,
        <QuestionBox
          inputFields={
            Array [
              <TextField
                additionalClasses={
                  Array [
                    "formTextField",
                  ]
                }
                attributes={
                  Object {
                    "test-id": "town",
                  }
                }
                label="Town"
                onChange={[Function]}
                value=""
              />,
            ]
          }
          text=""
        />,
        <QuestionBox
          inputFields={
            Array [
              <TextField
                additionalClasses={
                  Array [
                    "formTextField",
                  ]
                }
                attributes={
                  Object {
                    "test-id": "county",
                  }
                }
                label="County"
                onChange={[Function]}
                value=""
              />,
            ]
          }
          text=""
        />,
        <QuestionBox
          inputFields={
            Array [
              <TextField
                additionalClasses={
                  Array [
                    "formTextField",
                  ]
                }
                attributes={
                  Object {
                    "test-id": "postcode",
                  }
                }
                label="Postcode"
                onChange={[Function]}
                value=""
              />,
            ]
          }
          text=""
        />,
      ]
    }
    visible={false}
  />
</div>
`;
